tests/integration/security/jwt_test.go

```diff
@@ -62,335 +61,284 @@ func TestRequestAuthentication(t *testing.T) {
+					},
+					{
+						Name:   "invalid-jwks-no-token-noauthz",
+						Config: "invalid-jwks",
+						CallOpts: echo.CallOptions{
+							PortName: "http",
+							Scheme:   scheme.HTTP,
+							Count:    callCount,
+						},
+						ExpectResponseCode: response.StatusCodeOK,
+					},
+				}
+				for _, c := range testCases {
+					echotest.New(t, apps.All).
+						SetupForPair(func(t framework.TestContext, src, dst echo.Instances) error {
 							if c.Config != "" {
-								policy := tmpl.EvaluateOrFail(t,
-									file.AsStringOrFail(t, fmt.Sprintf("testdata/requestauthn/%s.yaml.tmpl", c.Config)), namespaceTmpl)
-								t.Config().ApplyYAMLOrFail(t, ns.Name(), policy)
-								util.WaitForConfig(t, policy, ns)
+								policy := yml.MustApplyNamespace(t, tmpl.MustEvaluate(
+									file.AsStringOrFail(t, fmt.Sprintf("testdata/requestauthn/%s.yaml.tmpl", c.Config)),
+									map[string]string{
+										"Namespace": ns.Name(),
+										"dst":       dst[0].Config().Service,
+									},
+								), ns.Name())
+								return t.Config().ApplyYAML(ns.Name(), policy)
 							}
-
-							retry.UntilSuccessOrFail(t, c.CheckAuthn, echo.DefaultCallRetryOptions()...)
+							return nil
+						}).
+						From(
+							// TODO(JimmyCYJ): enable VM and fix valid-token-forward-remote-jwks and invalid_aud.
+							filters(t, ns.Name(), true)...).
+						ConditionallyTo(echotest.ReachableDestinations).
+						To(filters(t, ns.Name(), false)...).
+						Run(func(t framework.TestContext, src echo.Instance, dest echo.Instances) {
+							t.NewSubTest(c.Name).Run(func(t framework.TestContext) {
+								c.CallOpts.Target = dest[0]
+								c.DestClusters = dest.Match(echo.InCluster(src.Config().Cluster)).Clusters()
+								c.CallOpts.Validator = echo.And(echo.ValidatorFunc(c.CheckAuthn))
+								src.CallWithRetryOrFail(t, c.CallOpts, echo.DefaultCallRetryOptions()...)
+							})
 						})
-					}
-				})
-			}
+				}
+			})
 		})
 }
 
+// filters returns a set of filters to only select injected workloads. VM is skipped
+// if skipVM is true.
+// TODO(JimmyCYJ): Simplify this function as a single filter.
+func filters(t framework.TestContext, ns string, skipVM bool) []echotest.Filter {
+	rt := []echotest.Filter{
+		echotest.SingleSimplePodServiceAndAllSpecial(),
+		echotest.Not(func(instances echo.Instances) echo.Instances { return instances.Match(echo.IsHeadless()) }),
+		echotest.Not(func(instances echo.Instances) echo.Instances { return instances.Match(echo.IsNaked()) }),
+		echotest.Not(func(instances echo.Instances) echo.Instances { return instances.Match(echo.IsExternal()) }),
+		echotest.Not(func(instances echo.Instances) echo.Instances { return instances.Match(util.IsMultiversion()) }),
+		func(instances echo.Instances) echo.Instances { return instances.Match(echo.Namespace(ns)) },
+		func(instances echo.Instances) echo.Instances {
+			return instances.Match(echo.InCluster(t.Clusters().Default()))
```

Comment:
1. This will not stop us from _attempting_ to send traffic there.
2. If this has to do with east-west gateway stripping headers, try to make this a combination filter, that results in same-network endpoints (not just same cluster)
3. Please create an issue/TODO linking that issue so we don't lose track of this. I spent time this morning enumerating the testing gaps in multicluster and there are plenty.

Reply:
This test does not generate cross-cluster traffic, I added a TODO to cover cross-cluster communication.